### ISSUE TITLE
LineageOS OTA update wipes out magisk if /data is encrypted

### ISSUE THREAD

**Issue by TjrGithub** [topjohnwu/Magisk#1500]
The LineageOS OTA update app uninstalls Magisk. While it does make TWRP run the Magisk addon.d script, it does not decrypt the userdata partition. The addon-script then cannot find something and refuses to reflash Magisk.

The current workaround is to hold the hardware keys in the first TWRP boot, so that the reboot goes directly into TWRP again. Then, I can flash the magisk zip.

Versions:
Magisk and Manager as of 2019-05-10
Xiaomi Redmi Note 4 (Mido) with Microg https://download.lineage.microg.org/mido/lineage-15.1-20190508-microG-mido.zip

SEe also #175

**Comment by topjohnwu** [comment]
Nothing I can fix

**Comment by TjrGithub** [comment]
Sure you can fix that. Just one idea needs to be workable:

* Magisk addon.d script could look at the external SD card, which usually isn't encrypted.

* You could store the .zip in the same place as the addon.d, since it gets copied over by OpenRecoveryScript. (Or only the boot partition part of the zip.)

* The addon.d script could look for the .zip in the recovery partition. There are 434 Megabytes of free space (difference between Mido recovery parition and twrp image size).

* Magisk Manager could somehow ask the Lineage Updater app to trigger it, and unpack the flashable .zip into the same area as the OS update.

I'm sure there are lots of other possibilities.

**Comment by ARitz-Cracker** [comment]
I'm commenting here so the next person who searches for this issue can find a quick solution.

Obligatory: It worked flawlessly for me but might not for you.

The problem is that the magisk post-update script needs the data partition to be decrypted to do what it needs to do. The LineageOS Recovery doesn't give you the option to decrypt your data during an update, but TWRP does. TWRP appears to be compatible with the User-space LineageOS updater, so once you download the update, (and remember to disable "updating" the recovery environment in the updater preferences) you can hit "install", your phone will reboot to TWRP, which will automatically install the update after you decrypt your data, then magisk will also re-install itself automatically.

tldr: use TWRP instead of LineageOS stock recovery if you want to keep magisk after updating the OS.

**Comment by CoelacanthusHex** [comment]
> I'm commenting here so the next person who searches for this issue can find a quick solution.
> 
> Obligatory: It worked flawlessly for me but might not for you.
> 
> The problem is that the magisk post-update script needs the data partition to be decrypted to do what it needs to do. The LineageOS Recovery doesn't give you the option to decrypt your data during an update, but TWRP does. TWRP appears to be compatible with the User-space LineageOS updater, so once you download the update, (and remember to disable "updating" the recovery environment in the updater preferences) you can hit "install", your phone will reboot to TWRP, which will automatically install the update after you decrypt your data, then magisk will also re-install itself automatically.
> 
> tldr: use TWRP instead of LineageOS stock recovery if you want to keep magisk after updating the OS.

Thx

**Comment by joaocandre** [comment]
> I'm commenting here so the next person who searches for this issue can find a quick solution.
> 
> Obligatory: It worked flawlessly for me but might not for you.
> 
> The problem is that the magisk post-update script needs the data partition to be decrypted to do what it needs to do. The LineageOS Recovery doesn't give you the option to decrypt your data during an update, but TWRP does. TWRP appears to be compatible with the User-space LineageOS updater, so once you download the update, (and remember to disable "updating" the recovery environment in the updater preferences) you can hit "install", your phone will reboot to TWRP, which will automatically install the update after you decrypt your data, then magisk will also re-install itself automatically.
> 
> tldr: use TWRP instead of LineageOS stock recovery if you want to keep magisk after updating the OS.

For reference, using TWRP recovery I still got Magisk wiped on Lineage w/ MicroG 18.1 after an OTA update. Not sure why it happened, considering I needed to input the screen unlock pattern for it to decrypt the partitions before the update.
